src/drivers/virtio/virtqueue/split.rs

```diff
@@ -45,40 +48,113 @@ impl Descriptor {
+	fn ring_ptr<A: volatile::access::Access>(
+		volatile_self: VolatilePtr<'_, Self, A>,
+	) -> VolatilePtr<'_, [MaybeUninit<le16>], A> {
+		let ring_and_event_ptr = map_field!(volatile_self.ring_and_event);
+		ring_and_event_ptr.split_at(ring_and_event_ptr.len()).0
+	}
+
+	fn event_ptr<A: volatile::access::Access>(
+		volatile_self: VolatilePtr<'_, Self, A>,
+	) -> VolatilePtr<'_, MaybeUninit<le16>, A> {
+		let ring_and_event_ptr = map_field!(volatile_self.ring_and_event);
+		ring_and_event_ptr.index(ring_and_event_ptr.len() - 1)
+	}
 }
 
-struct UsedRing {
-	flags: &'static mut little_endian::U16,
-	index: *mut little_endian::U16,
-	ring: &'static mut [UsedElem],
-	event: &'static mut little_endian::U16,
+// The elements of the unsized field and the last field are not of the same type.
+// For this reason, the field stores raw bytes and we have typed accessors.
+type UsedRing = GenericRing<[u8]>;
+
+// Used ring is not supposed to be modified by the driver. Thus, we only have _ref methods (and not _mut methods).
+impl UsedRing {
+	fn ring_ptr<A: volatile::access::Access>(
+		volatile_self: VolatilePtr<'_, Self, A>,
+	) -> VolatilePtr<'_, [UsedElem], A> {
+		let ring_and_event_ptr = map_field!(volatile_self.ring_and_event);
+		let ring_len = (ring_and_event_ptr.len() - size_of::<le16>()) / size_of::<UsedElem>();
+
+		unsafe {
+			ring_and_event_ptr.map(|ring_and_event_ptr| {
+				NonNull::slice_from_raw_parts(ring_and_event_ptr.cast::<UsedElem>(), ring_len)
+			})
+		}
+	}
+
+	fn event_ptr<A: volatile::access::Access>(
+		volatile_self: VolatilePtr<'_, Self, A>,
+	) -> VolatilePtr<'_, le16, A> {
+		let ring_and_event_ptr = map_field!(volatile_self.ring_and_event);
+		let ring_and_event_len = ring_and_event_ptr.len();
+		let event_bytes_ptr = ring_and_event_ptr
+			.split_at(ring_and_event_len - size_of::<le16>())
+			.1;
+
+		unsafe { event_bytes_ptr.map(|event_bytes| event_bytes.cast::<le16>()) }
+	}
 }
 
 #[repr(C)]
 #[derive(Copy, Clone)]
 struct UsedElem {
-	id: little_endian::U32,
-	len: little_endian::U32,
+	id: le32,
+	len: le32,
 }
 
 struct DescrRing {
 	read_idx: u16,
-	descr_table: DescrTable,
-	ref_ring: Box<[Option<Box<TransferToken>>]>,
-	avail_ring: AvailRing,
-	used_ring: UsedRing,
+	token_ring: Box<[Option<Box<TransferToken>>]>,
+
+	// The following fields should not be used directly, as they do not ensure volatile access.
+	#[deprecated = "_*_cell fields are only for keeping the objects around. All access should be done through the respective *_ref methods."]
+	_descr_table_cell: Box<UnsafeCell<[MaybeUninit<Descriptor>]>, DeviceAlloc>,
+	#[deprecated = "_*_cell fields are only for keeping the objects around. All access should be done through the respective *_ref methods."]
+	_avail_ring_cell: Box<UnsafeCell<AvailRing>, DeviceAlloc>,
+	#[deprecated = "_*_cell fields are only for keeping the objects around. All access should be done through the respective *_ref methods."]
+	_used_ring_cell: Box<UnsafeCell<UsedRing>, DeviceAlloc>,
```

Comment:
I think we don't benefit from the noise of having `_` prefixes and deprecation attributes. Any access would have to be `unsafe` anyway, with the invariant being that access has to be volatile, which is conveniently provided by methods of the same name. The `_` prefix should only be used when we want to silence dead-code warnings but need the value around for side effects. Deprecation attributes should be used for things that are slated for removal, but these are not. Instead, it is very valid to use these fields when performing volatile accesses, which the wrapper method guarantees. We should change these attributes to be safety comments. Something like this:

```suggestion
    /// Descriptor Tables
    ///
    /// # Safety
    ///
    /// These tables may only be accessed via volatile operations.
    /// See the corresponding method for a safe wrapper.
	descr_table: Box<UnsafeCell<[MaybeUninit<Descriptor>]>, DeviceAlloc>,
	
	/// ...
	avail_ring: Box<UnsafeCell<AvailRing>, DeviceAlloc>,
	used_ring: Box<UnsafeCell<UsedRing>, DeviceAlloc>,
```